### PR TITLE
Bound pallet-offences

### PR DESCRIPTION
**Pull request by ggwpez**
Bound the `offences` pallet.  
Changes:
- Introduce global `MaxOffenders` and `MaxReportes` constants with both value 100. This is necessary since the staking and offences traits would otherwise all need to be parameterized. Grandpa and BABE both only use at most 1 for these values, so 100 should be enough.
- Introduce `MaxSameKindReportsEncodedLen` and `MaxOpaqueTimeSlotEncodedLen` to bound the storage of encoded values. There is a test which tests that these constants can encode the respective values, it would be important to run that test against the real runtime config, not just the mock.
- Introduce `MaxConcurrentReportsPerTime` and `MaxSameKindReports`. These may or may not be necessary, I am not sure. Maybe they can be replaced with instances of `MaxOffenders/MaxReporters`. 
- Change the naming of `Exposure/ExposureOf` and `ActiveExposure` to avoid a name-clash
- Replace a lot of `vec![]` with `Default::default()` to make future refactoring easier.

TODO:
- There is one `expect("todo")` where I was unsure about the vector length
- What about migrations?

Part of #8629

### CODE REVIEW

**Comment by KiChjang** [comment]
Would really like @kianenigma to review this when he comes back...

**Comment by ggwpez** [comment]
bot rebase

File: frame/babe/src/lib.rs [hunk 1]
```diff
@@ -826,10 +827,13 @@ impl<T: Config> Pallet<T> {
...
 			BabeEquivocationOffence { slot, validator_set_count, offender, session_index };
 
-		let reporters = match reporter {
+		let reporters: BoundedVec<T::AccountId, MaxReporters> = match reporter {
 			Some(id) => vec![id],
 			None => vec![],
-		};
+		}
+		.try_into()
+		.defensive_proof("MaxReporters must be at least 1;")
```
Comment (kianenigma) [inline frame/babe/src/lib.rs:835]: do we have `defensive_expect`? we should :D

File: frame/grandpa/src/lib.rs [hunk 2]
```diff
@@ -568,9 +568,16 @@ impl<T: Config> Pallet<T> {
 			return Err(Error::<T>::InvalidEquivocationProof.into())
 		}
 
+		let reporters: BoundedVec<T::AccountId, MaxReporters> = match reporter {
+			Some(id) => vec![id],
+			None => vec![],
+		}
+		.try_into()
+		.expect("MaxReporters must be at least 1;");
```
Comment (kianenigma) [inline frame/grandpa/src/lib.rs:576]: where is this ensured? Should be in `integrity_checks`

File: frame/offences/src/lib.rs [hunk 3]
```diff
@@ -26,42 +26,68 @@ pub mod migration;
...
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
 		/// Full identification of the validator.
-		type IdentificationTuple: Parameter;
+		type IdentificationTuple: Parameter + MaxEncodedLen;
 		/// A handler called for every offence report.
 		type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple, Weight>;
+
+		/// Maximum number of reports of the same kind that happened at a specific time slot.
+		#[pallet::constant]
```
Comment (kianenigma) [inline frame/offences/src/lib.rs:74]: Do all of these really need to be in metadata? harmless, but I wonder if you had thought about each.

File: frame/staking/src/lib.rs [hunk 4]
```diff
@@ -696,21 +702,45 @@ pub struct IndividualExposure<AccountId, Balance: HasCompact> {
...
+	TypeInfo,
+	MaxEncodedLen,
+)]
+#[codec(mel_bound())]
+#[scale_info(skip_type_params(MaxNominations))]
+pub struct Exposure<AccountId, Balance, MaxNominations>
+where
+	AccountId: Debug + Encode + Decode + MaxEncodedLen + Clone + Eq + PartialEq,
+	Balance: Debug + HasCompact + Encode + Decode + MaxEncodedLen + Clone + Eq + PartialEq,
+	MaxNominations: Get<u32>,
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:723]: Not a good name -- we use this term to determine the number of nominations that a single nominator can have. i.e. 16. 

I suggest using `MaxBackers` here instead.

File: frame/staking/src/lib.rs [hunk 5]
```diff
@@ -696,21 +702,45 @@ pub struct IndividualExposure<AccountId, Balance: HasCompact> {
...
+{
 	/// The total balance backing this validator.
 	#[codec(compact)]
 	pub total: Balance,
 	/// The validator's own stash that is exposed.
 	#[codec(compact)]
 	pub own: Balance,
 	/// The portions of nominators stashes that are exposed.
-	pub others: Vec<IndividualExposure<AccountId, Balance>>,
+	pub others: BoundedVec<IndividualExposure<AccountId, Balance>, MaxNominations>,
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:732]: This is a great direction that you started here: we certainly want to bound this. Problem is that there is nothing in the `ElectionProvider` trait explicitly guarantees that this is the case. 

What we want to do is:

1. Staking tells `ElectionProvider` how many `MaxBackers` it accepts. 
2. Staking miner respects this when generating solution, and automatically sorts the backings of each validator, and throws away the smaller ones. (currently we area doing onchain, which is horrible)
3. `ElectionProvider` checks this for submitted solutions and ensures it is the case.
4.  Staking can just do a defensive truncate when it receives a solution from `ElectionProvider`. 
5. This new `MaxBackers` should also replace `MaxRewardableNominatorsPerValidator`, and entirely replace the notion of oversubscribed validators. 

If interested in doing more steps in this path, I will be happy to do a better writeup about it.

File: frame/staking/src/lib.rs [hunk 6]
```diff
@@ -895,17 +925,19 @@ impl<T: Config> Convert<T::AccountId, Option<T::AccountId>> for StashOf<T> {
...
+pub type ExposureOf<T> =
+	Exposure<<T as frame_system::Config>::AccountId, BalanceOf<T>, <T as Config>::MaxNominations>;
+
 /// A typed conversion from stash account ID to the active exposure of nominators
 /// on that account.
 ///
 /// Active exposure is the exposure of the validator set currently validating, i.e. in
 /// `active_era`. It can differ from the latest planned exposure in `current_era`.
-pub struct ExposureOf<T>(sp_std::marker::PhantomData<T>);
+pub struct ActiveExposure<T>(sp_std::marker::PhantomData<T>);
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:937]: better name now, but still not quite happy with it.

File: frame/staking/src/pallet/impls.rs [hunk 7]
```diff
@@ -553,10 +552,10 @@ impl<T: Config> Pallet<T> {
 						total = total.saturating_add(stake);
 					});
 
-				let exposure = Exposure { own, others, total };
+				let exposure = Exposure { own, others: others.try_into().expect("TODO"), total };
```
Comment (kianenigma) [inline frame/staking/src/pallet/impls.rs:555]: This is exactly https://github.com/paritytech/substrate/pull/11585/files#r925580489

**Pull request closed without merging**